pkg/utils/kubernetes/status.go

```diff
@@ -0,0 +1,329 @@
+	for _, condition := range pod.Status.Conditions {
+		if condition.Type == corev1.PodReady && (condition.Reason == "PodCompleted" || condition.Status == corev1.ConditionTrue) {
+			return StatusReady, nil
+		}
+	}
+	return StatusInProgress, nil
+}
+
+// DeploymentStatus returns the status of a Deployment.
+func DeploymentStatus(dp *appsv1.Deployment) (StatusType, error) {
+	replicaFailure := false
+	progressing := false
+	available := false
+
+	for _, condition := range dp.Status.Conditions {
+		switch condition.Type {
+		case appsv1.DeploymentProgressing:
+			if condition.Status == corev1.ConditionTrue && condition.Reason == "NewReplicaSetAvailable" {
+				progressing = true
+			}
+		case appsv1.DeploymentAvailable:
+			if condition.Status == corev1.ConditionTrue {
+				available = true
+			}
+		case appsv1.DeploymentReplicaFailure:
+			if condition.Status == corev1.ConditionTrue {
+				replicaFailure = true
+				break
+			}
+		}
+	}
+
+	if dp.Status.ObservedGeneration == dp.Generation &&
+		dp.Status.Replicas == *dp.Spec.Replicas &&
+		dp.Status.ReadyReplicas == *dp.Spec.Replicas &&
+		dp.Status.AvailableReplicas >= *dp.Spec.Replicas &&
+		dp.Status.Conditions != nil && len(dp.Status.Conditions) > 0 &&
+		(progressing || available) && !replicaFailure {
+
+		return StatusReady, nil
+	}
+
+	return StatusInProgress, nil
+}
+
+// ReplicasetStatus returns the status of a ReplicasetStatus.
+func ReplicasetStatus(rs *appsv1.ReplicaSet) (StatusType, error) {
+	replicaFailure := false
+	for _, condition := range rs.Status.Conditions {
+		switch condition.Type {
+		case appsv1.ReplicaSetReplicaFailure:
+			if condition.Status == corev1.ConditionTrue {
+				replicaFailure = true
+				break
+			}
+		}
+	}
+	if rs.Status.ObservedGeneration == rs.Generation &&
+		rs.Status.Replicas == *rs.Spec.Replicas &&
+		rs.Status.ReadyReplicas >= *rs.Spec.Replicas &&
+		rs.Status.AvailableReplicas == *rs.Spec.Replicas &&
+		!replicaFailure {
+
+		return StatusReady, nil
+	}
+	return StatusInProgress, nil
+}
+
+// StatefulSetStatus returns the status of a StatefulSet.
+func StatefulSetStatus(sts *appsv1.StatefulSet) (StatusType, error) {
+	if sts.Status.ObservedGeneration == sts.Generation &&
+		sts.Status.Replicas == *sts.Spec.Replicas &&
+		sts.Status.ReadyReplicas == *sts.Spec.Replicas &&
+		sts.Status.CurrentReplicas == *sts.Spec.Replicas {
+
+		return StatusReady, nil
+	}
+	return StatusInProgress, nil
+}
+
+// DaemonSetStatus returns the status of a DaemonSet.
+func DaemonSetStatus(ds *appsv1.DaemonSet) (StatusType, error) {
+	if ds.Status.ObservedGeneration == ds.Generation &&
+		ds.Status.DesiredNumberScheduled == ds.Status.NumberAvailable &&
+		ds.Status.DesiredNumberScheduled == ds.Status.NumberReady {
+
+		return StatusReady, nil
+	}
+	return StatusInProgress, nil
+}
+
+// ReplicationControllerStatus returns the status of a ReplicationController.
+func ReplicationControllerStatus(rc *corev1.ReplicationController) (StatusType, error) {
```

Comment:
maybe here it would make sense to not only report the status but also why it is in that status.
as for example gardener does it in https://github.com/gardener/gardener/blob/master/pkg/utils/kubernetes/health/health.go#L90 
It returns an error with the concrete issue

Reply:
yes will do that